service/history/memory_scheduled_queue.go

```diff
@@ -0,0 +1,314 @@
+			return
+		default:
+		}
+
+		select {
+		case <-q.shutdownCh:
+			return
+		case newTask := <-q.newTaskCh:
+			var nextTimeoutTime time.Time
+			if !q.timeoutQueue.IsEmpty() {
+				nextTimeoutTime = q.timeoutQueue.Peek().ScheduledTime()
+			}
+			q.timeoutQueue.Add(newTask)
+			// If there is no timer set OR new time is earlier than the current one, then timer needs to be set to the new time.
+			if nextTimeoutTime.IsZero() || newTask.ScheduledTime().Before(nextTimeoutTime) {
+				// But before reset if timer is there it needs to be stopped.
+				if !nextTimeoutTime.IsZero() {
+					if !q.nextTimeoutTimer.Stop() {
+						// Drain timer channel to prevent timer firing.
+						<-q.nextTimeoutTimer.C
+					}
+				}
+				q.nextTimeoutTimer.Reset(newTask.ScheduledTime().Sub(q.timeSource.Now()))
+			}
+			q.metricsHandler.Counter(metrics.NewTimerNotifyCounter.GetMetricName()).Record(1)
+		case <-q.nextTimeoutTimer.C:
+			taskToExecute := q.timeoutQueue.Remove()
+			// Skip tasks which are already canceled. Majority of the tasks in the queue should be cancelled already.
+			nextTask := q.purgeCanceledTasks()
+			if nextTask != nil {
+				q.nextTimeoutTimer.Reset(nextTask.ScheduledTime().Sub(q.timeSource.Now()))
+			}
+			// If current taskToExecute is also cancelled don't submit it to scheduler.
+			if taskToExecute.State() == ctasks.TaskStateCancelled {
+				continue
+			}
+
+			taskToExecute.SetExecutorFunc(q.executeTask)
+			q.scheduler.Submit(taskToExecute)
+		}
+	}
+}
+func (q *MemoryScheduledQueue) purgeCanceledTasks() (nextTask *tasks.SpeculativeWorkflowTaskTimeoutTask) {
+	if !q.timeoutQueue.IsEmpty() {
+		nextTask = q.timeoutQueue.Peek()
+	}
+	for nextTask != nil {
+		if nextTask.State() != ctasks.TaskStateCancelled {
+			return
+		}
+		// Remove canceled task from queue.
+		q.timeoutQueue.Remove()
+		if q.timeoutQueue.IsEmpty() {
+			return nil
+		}
+		nextTask = q.timeoutQueue.Peek()
+	}
+	return
+}
+
+func (q *MemoryScheduledQueue) executeTask(t *tasks.SpeculativeWorkflowTaskTimeoutTask) (retError error) {
+	// This function is mostly copied form timerQueueActiveTaskExecutor.executeWorkflowTaskTimeoutTask()
+	// After timerQueueActiveTaskExecutor is moved out of history root package, it must be used here instead.
+
+	ctx, cancel := context.WithTimeout(context.Background(), taskTimeout)
+	defer cancel()
+
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, q.workflowCache, t)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
+	// Check task state again, after acquiring workflow lock.
+	if t.State() == ctasks.TaskStateCancelled {
+		return nil
+	}
+
+	mutableState, err := loadMutableStateForTimerTask(ctx, weContext, t, q.metricsHandler, q.logger)
+	if err != nil {
+		return err
+	}
+	if mutableState == nil || !mutableState.IsWorkflowExecutionRunning() {
+		return nil
+	}
+
+	// This will clear task pointer in mutable state.
+	mutableState.CancelSpeculativeWorkflowTaskTimeoutTask()
```

Comment:
if the first attempt fails, then there will be no second attempt since it's already canceled?

Reply:
actually, whenever this task fails, mutable state will be cleared and there will be no speculative workflow task after reload? So this task should never retry?